[raageth] iP

raageth opened the pull request:
# Heading

## Bullet List
- First
- Second
  - Nested
- Third

## Numbered List
1. First
2. Second
3. Third

## Code Block
```java
public class HelloWorld {
    public static void main(String[] args) {
        System.out.println("Hello, World!");
    }
}
```

## Task List
- [ ] Task 1
- [x] Task 2
- [ ] Task 3

## Emoji
💯 

## Blockquote
> Blockquote

## Hyperlink
[repo](https://github.com/raageth/ip/tree/master)

## Inline Code
```a = 1```

## Text Formatting
**bold** *italic* ~~strikethrough~~

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

raageth committed:
Replace Duke with Darwin

raageth committed:
Add echo functionality

raageth committed:
Add list functionality

raageth committed:
Fix bug in chat logic

raageth committed:
Add ability to change status of tasks

raageth committed:
Fix task index formatting issue

raageth committed:
Add multi task tracking

raageth committed:
Add automated ui testing

raageth committed:
Handle incorrect inputs

raageth committed:
Support deletion of tasks

raageth committed:
Add javadocs

raageth committed:
Add ability to save tasks to hard drive

raageth committed:
Merge branch 'branch-level-7'

raageth committed:
Refactor TaskFactory switch statements

raageth committed:
Add task parsers to handle input

raageth committed:
Merge branch 'branch-Level-8'

raageth committed:
Add commands, input parser

raageth committed:
Update TaskManager save logic

raageth committed:
Stop tracking task storage

raageth committed:
Merge branch 'add-gradle-support'

raageth committed:
Update build file

raageth committed:
Add tests

raageth committed:
Add jar

raageth committed:
Move packages to darwin package

raageth committed:
Move packages to darwin package

raageth committed:
Add javadoc

raageth committed:
Ensure code is compliant to standard

raageth committed:
Add ability to find tasks

raageth committed:
Merge branch 'branch-A-JavaDoc'

raageth committed:
Merge branch 'branch-A-CodingStandard'

raageth committed:
Merge branch 'branch-Level-9'

raageth committed:
Convert storage path to field

raageth committed:
Clean code formatting

raageth committed:
Add GUI to chatbot

raageth committed:
Merge branch 'branch-Level-10'

linedoestrolling commented on src/main/java/darwin/exception/IllegalTaskTypeException.java:
I like all these specific exceptions for the different things that could go wrong.

linedoestrolling commented on src/main/java/darwin/task/TaskType.java:
I like the use of enum here to have your own methods and to fix what kind of tasks are accepted, which can then be used in many other places in your code.

linedoestrolling commented on src/main/java/darwin/parser/TaskFactory.java:
```diff
@@ -0,0 +1,47 @@
+ * TaskFactory class to create Task instances from user input or database string.
+ */
+public class TaskFactory {
+    private static final HashMap<TaskType, TaskParser> TASK_TYPE_PARSER_MAP = new HashMap<>() {{
```
Good use of a hashmap here to map your tasktype to taskeparser.

linedoestrolling commented on src/main/java/darwin/parser/InputParser.java:
```diff
@@ -0,0 +1,92 @@
+        String args = parts.length == 2 ? parts[1] : InputParser.EMPTY_ARGS;
+
+        switch (cmd) {
+        case ExitCommand.CMD_WORD -> {
```
Since your case bodies are only 1 line long, maybe you could use the default syntax given in the coding standard?
Cases would not use the arrow and braces but instead only have a colon.

linedoestrolling reviewed (commented):
Overall, great factoring of code into relevant classes, made looking through the code easy and simple to understand.

Bonzzz3 commented on src/main/java/darwin/command/DeleteCommand.java:
```diff
@@ -0,0 +1,37 @@
+public class DeleteCommand extends Command {
+    public static final String CMD_WORD = "delete";
+    private static final String DELETE_TASK_MSG = "Noted. I've removed this task:";
+    private static final String TASK_COUNT_MSG = "Now you have %d tasks in the list.";
```
Good naming for constants!

Bonzzz3 commented on src/main/java/darwin/parser/EventParser.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    @Override
+    public Task parseFromDb(String taskStr) throws IllegalTaskArgumentException {
```
Db might be a little bit vague imo. Others might think of it as decibel.

Bonzzz3 commented on src/main/java/darwin/storage/DbManager.java:
```diff
@@ -0,0 +1,69 @@
+        this.path = path;
+    }
+
+    private void ensureFileExists(String filePath) {
```
Good naming for method (using verbs).

Bonzzz3 commented on src/main/java/darwin/command/ErrorCommand.java:
```diff
@@ -0,0 +1,19 @@
+
+public class ErrorCommand extends Command {
+    public static final String CMD_WORD = "";
+    private DarwinException e;
```
Maybe e could be written as error instead.

Bonzzz3 commented on src/main/java/darwin/ui/Ui.java:
```diff
@@ -0,0 +1,43 @@
+ */
+public class Ui {
+    private final Scanner scanner = new Scanner(System.in);
+    private String lastResponse;
```
Clear and good naming for string (as noun).

Bonzzz3 commented on src/main/java/darwin/task/TaskType.java:
```diff
@@ -0,0 +1,55 @@
+/**
+ * Enum to represent the different types of tasks.
+ */
+public enum TaskType {
```
Good use of enum and naming of its constants.

Bonzzz3 reviewed (commented):
Good naming. No mistakes in naming found :)

MiloTruck commented on src/main/java/darwin/gui/Main.java:
```diff
@@ -0,0 +1,29 @@
+    @Override
+    public void start(Stage stage) {
+        try {
+            stage.setMinHeight(220);
+            stage.setMinWidth(417);
+            FXMLLoader fxmlLoader = new FXMLLoader(Main.class.getResource("/view/MainWindow.fxml"));
+            AnchorPane ap = fxmlLoader.load();
+            Scene scene = new Scene(ap);
+            stage.setScene(scene);
+            fxmlLoader.<MainWindow>getController().setDarwin(new Darwin("./tasks.txt"));
+            stage.show();
```
Would be good to space out your code based on their logical parts, instead of one huge chunk of code.

MiloTruck commented on src/main/java/darwin/parser/TaskParser.java:
```diff
@@ -0,0 +1,37 @@
+    Task parseFromDb(String taskStr) throws IllegalTaskTypeException, IllegalTaskArgumentException;
+
+    default void handleTaskStatus(Task task, String taskStatus) throws IllegalTaskArgumentException {
+        switch (taskStatus) {
+        case "1" -> task.markDone();
+        case "0" -> task.unmarkDone();
+        default -> throw new IllegalTaskArgumentException("");
+        }
```
This doesn't follow the switch statement form shown in the [coding standard](https://se-education.org/guides/conventions/java/intermediate.html#layout):

```java
switch (condition) {
case ABC:
    statements;
    // Fallthrough
case DEF:
    statements;
    break;
case XYZ:
    statements;
    break;
default:
    statements;
    break;
}
```

MiloTruck commented on src/main/java/darwin/task/TaskType.java:
```diff
@@ -0,0 +1,55 @@
+        return this.name;
+    }
+
+    public String getSymbol() { return this.symbol;}
```
Should use Egyptian style braces here:

```java
public String getSymbol() {
    return this.symbol;
}
```

MiloTruck commented on src/main/java/darwin/command/AddTaskCommand.java:
```diff
@@ -0,0 +1,31 @@
+    @Override
+    public void execute(TaskManager taskManager, Ui ui) {
+        taskManager.addTask(this.task);
+        // TODO: reduce repeated task count msg code in delete command
+        ui.send(String.format("%s\n    %s\n%s", ADD_TASK_MSG, task.getTaskInfo(),
+                String.format(TASK_COUNT_MSG, taskManager.getTaskCount())));
```
Readability can be improved here by formatting the function call as such:

```java
ui.send(String.format(
    "%s\n    %s\n%s", 
    ADD_TASK_MSG, 
    task.getTaskInfo(),
    String.format(TASK_COUNT_MSG, taskManager.getTaskCount())
));
```

MiloTruck commented on src/main/java/darwin/command/DeleteCommand.java:
```diff
@@ -0,0 +1,37 @@
+    @Override
+    public void execute(TaskManager taskManager, Ui ui) throws IllegalTaskNumberException {
+        Task task = taskManager.deleteTask(this.taskIdx);
+        ui.send(String.format("%s\n    %s\n%s", DELETE_TASK_MSG, task.getTaskInfo(),
+                String.format(TASK_COUNT_MSG, taskManager.getTaskCount())));
```
Readability here can be improved:

```java
ui.send(String.format(
    "%s\n    %s\n%s", 
    DELETE_TASK_MSG, 
    task.getTaskInfo(),
    String.format(TASK_COUNT_MSG, taskManager.getTaskCount())
));
```

MiloTruck reviewed (commented):
Lgtm. Just some suggestions to follow the coding standard.

raageth committed:
Potential code readability issues and deviation from coding standards

Reduces readability as code does not follow conventions.

Remove magic literals and implement wrapping for string formatting.

raageth committed:
Add assertions

raageth committed:
Space out code in gui class Main

raageth committed:
Merge pull request #1 from raageth/branch-A-Assertions

Add assertions

raageth committed:
Merge pull request #2 from raageth/branch-A-CodeQuality

Improve code quality

raageth committed:
Allow users to use a symbol for the task commands and fix minor bugs

raageth committed:
Merge branch 'branch-BCD-Extension'

raageth committed:
Clean up code and add javadocs

raageth committed:
Update readme

raageth committed:
Fix readme

raageth committed:
Update readme and docs

raageth committed:
Update readme

raageth committed:
Fix readme formatting

raageth committed:
Remove background
